File: Stripe/StripeiOSTests/STPBankAccountTest.m
//
//  STPBankAccountTest.m
//  Stripe
//
//  Created by Charles Scalesse on 10/2/14.
//
//

@import XCTest;



#import "STPTestUtils.h"

@interface STPBankAccount ()

+ (STPBankAccountStatus)statusFromString:(NSString *)string;
+ (NSString *)stringFromStatus:(STPBankAccountStatus)status;

- (void)setLast4:(NSString *)last4;

@end

@interface STPBankAccountTest : XCTestCase

@end

@implementation STPBankAccountTest

#pragma mark - STPBankAccountStatus Tests

- (void)testStatusFromString {
    XCTAssertEqual([STPBankAccount statusFromString:@"new"], STPBankAccountStatusNew);
    XCTAssertEqual([STPBankAccount statusFromString:@"NEW"], STPBankAccountStatusNew);

    XCTAssertEqual([STPBankAccount statusFromString:@"validated"], STPBankAccountStatusValidated);
    XCTAssertEqual([STPBankAccount statusFromString:@"VALIDATED"], STPBankAccountStatusValidated);

    XCTAssertEqual([STPBankAccount statusFromString:@"verified"], STPBankAccountStatusVerified);
    XCTAssertEqual([STPBankAccount statusFromString:@"VERIFIED"], STPBankAccountStatusVerified);

    XCTAssertEqual([STPBankAccount statusFromString:@"verification_failed"], STPBankAccountStatusVerificationFailed);
    XCTAssertEqual([STPBankAccount statusFromString:@"VERIFICATION_FAILED"], STPBankAccountStatusVerificationFailed);

    XCTAssertEqual([STPBankAccount statusFromString:@"errored"], STPBankAccountStatusErrored);
    XCTAssertEqual([STPBankAccount statusFromString:@"ERRORED"], STPBankAccountStatusErrored);

    XCTAssertEqual([STPBankAccount statusFromString:@"garbage"], STPBankAccountStatusNew);
    XCTAssertEqual([STPBankAccount statusFromString:@"GARBAGE"], STPBankAccountStatusNew);
}

- (void)testStringFromStatus {
    NSArray<NSNumber *> *values = @[
                                    @(STPBankAccountStatusNew),
                                    @(STPBankAccountStatusValidated),
                                    @(STPBankAccountStatusVerified),
                                    @(STPBankAccountStatusVerificationFailed),
                                    @(STPBankAccountStatusErrored)
                                    ];

    for (NSNumber *statusNumber in values) {
        STPBankAccountStatus status = (STPBankAccountStatus)[statusNumber integerValue];
        NSString *string = [STPBankAccount stringFromStatus:status];

        switch (status) {
            case STPBankAccountStatusNew:
                XCTAssertEqualObjects(string, @"new");
                break;
            case STPBankAccountStatusValidated:
                XCTAssertEqualObjects(string, @"validated");
                break;
            case STPBankAccountStatusVerified:
                XCTAssertEqualObjects(string, @"verified");
                break;
            case STPBankAccountStatusVerificationFailed:
                XCTAssertEqualObjects(string, @"verification_failed");
                break;
            case STPBankAccountStatusErrored:
                XCTAssertEqualObjects(string, @"errored");
                break;
        }
    }
}

#pragma mark - Equality Tests

- (void)testBankAccountEquals {
    STPBankAccount *bankAccount1 = [STPBankAccount decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"BankAccount"]];
    STPBankAccount *bankAccount2 = [STPBankAccount decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"BankAccount"]];

    XCTAssertNotEqual(bankAccount1, bankAccount2);

    XCTAssertEqualObjects(bankAccount1, bankAccount1);
    XCTAssertEqualObjects(bankAccount1, bankAccount2);

    XCTAssertEqual(bankAccount1.hash, bankAccount1.hash);
    XCTAssertEqual(bankAccount1.hash, bankAccount2.hash);
}

#pragma mark - Description Tests

- (void)testDescription {
    STPBankAccount *bankAccount = [STPBankAccount decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"BankAccount"]];
    XCTAssert(bankAccount.description);
}

#pragma mark - STPAPIResponseDecodable Tests

- (void)testDecodedObjectFromAPIResponseRequiredFields {
    NSArray<NSString *> *requiredFields = @[
                                            @"id",
                                            @"last4",
                                            @"bank_name",
                                            @"country",
                                            @"currency",
                                            @"status",
                                            ];

    for (NSString *field in requiredFields) {
        NSMutableDictionary *response = [[STPTestUtils jsonNamed:@"BankAccount"] mutableCopy];
        [response removeObjectForKey:field];

        XCTAssertNil([STPBankAccount decodedObjectFromAPIResponse:response]);
    }

    XCTAssert([STPBankAccount decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"BankAccount"]]);
}

- (void)testDecodedObjectFromAPIResponseMapping {
    NSDictionary *response = [STPTestUtils jsonNamed:@"BankAccount"];
    STPBankAccount *bankAccount = [STPBankAccount decodedObjectFromAPIResponse:response];

    XCTAssertEqualObjects(bankAccount.stripeID, @"ba_1AZmya2eZvKYlo2CQzt7Fwnz");
    XCTAssertEqualObjects(bankAccount.accountHolderName, @"Jane Austen");
    XCTAssertEqual(bankAccount.accountHolderType, STPBankAccountHolderTypeIndividual);
    XCTAssertEqualObjects(bankAccount.bankName, @"STRIPE TEST BANK");
    XCTAssertEqualObjects(bankAccount.country, @"US");
    XCTAssertEqualObjects(bankAccount.currency, @"usd");
    XCTAssertEqualObjects(bankAccount.fingerprint, @"1JWtPxqbdX5Gamtc");
    XCTAssertEqualObjects(bankAccount.last4, @"6789");
    XCTAssertEqualObjects(bankAccount.routingNumber, @"110000000");
    XCTAssertEqual(bankAccount.status, STPBankAccountStatusNew);

    XCTAssertNotEqual(bankAccount.allResponseFields, response);
    XCTAssertEqualObjects(bankAccount.allResponseFields, response);
}

@end

File: Stripe/StripeiOSTests/STPBankAccountTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPBankAccountTest.swift
//  Stripe
//
//  Created by Charles Scalesse on 10/2/14.
//
//

import XCTest

class STPBankAccount {
    private class func status(from string: String?) -> STPBankAccountStatus {
    }

    private class func string(from status: STPBankAccountStatus) -> String? {
    }

    private func setLast4(_ last4: String?) {
    }
}

class STPBankAccountTest: XCTestCase {
    // MARK: - STPBankAccountStatus Tests

    func testStatusFromString() {
        XCTAssertEqual(Int(STPBankAccount.status(from: "new")), Int(STPBankAccountStatusNew))
        XCTAssertEqual(Int(STPBankAccount.status(from: "NEW")), Int(STPBankAccountStatusNew))

        XCTAssertEqual(Int(STPBankAccount.status(from: "validated")), Int(STPBankAccountStatusValidated))
        XCTAssertEqual(Int(STPBankAccount.status(from: "VALIDATED")), Int(STPBankAccountStatusValidated))

        XCTAssertEqual(Int(STPBankAccount.status(from: "verified")), Int(STPBankAccountStatusVerified))
        XCTAssertEqual(Int(STPBankAccount.status(from: "VERIFIED")), Int(STPBankAccountStatusVerified))

        XCTAssertEqual(Int(STPBankAccount.status(from: "verification_failed")), Int(STPBankAccountStatusVerificationFailed))
        XCTAssertEqual(Int(STPBankAccount.status(from: "VERIFICATION_FAILED")), Int(STPBankAccountStatusVerificationFailed))

        XCTAssertEqual(Int(STPBankAccount.status(from: "errored")), Int(STPBankAccountStatusErrored))
        XCTAssertEqual(Int(STPBankAccount.status(from: "ERRORED")), Int(STPBankAccountStatusErrored))

        XCTAssertEqual(Int(STPBankAccount.status(from: "garbage")), Int(STPBankAccountStatusNew))
        XCTAssertEqual(Int(STPBankAccount.status(from: "GARBAGE")), Int(STPBankAccountStatusNew))
    }

    func testStringFromStatus() {
        let values = [
            NSNumber(value: STPBankAccountStatusNew),
            NSNumber(value: STPBankAccountStatusValidated),
            NSNumber(value: STPBankAccountStatusVerified),
            NSNumber(value: STPBankAccountStatusVerificationFailed),
            NSNumber(value: STPBankAccountStatusErrored)
        ]

        for statusNumber in values {
            let status = statusNumber.intValue as? STPBankAccountStatus
            var string: String?
            if let status {
                string = STPBankAccount.string(from: status)
            }

            switch status {
            case STPBankAccountStatusNew:
                XCTAssertEqual(string, "new")
            case STPBankAccountStatusValidated:
                XCTAssertEqual(string, "validated")
            case STPBankAccountStatusVerified:
                XCTAssertEqual(string, "verified")
            case STPBankAccountStatusVerificationFailed:
                XCTAssertEqual(string, "verification_failed")
            case STPBankAccountStatusErrored:
                XCTAssertEqual(string, "errored")
            default:
                break
            }
        }
    }

    // MARK: - Equality Tests

    func testBankAccountEquals() {
        let bankAccount1 = STPBankAccount.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("BankAccount"))
        let bankAccount2 = STPBankAccount.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("BankAccount"))

        XCTAssertNotEqual(bankAccount1, bankAccount2)

        XCTAssertEqual(bankAccount1, bankAccount1)
        XCTAssertEqual(bankAccount1, bankAccount2)

        XCTAssertEqual(bankAccount1?.hash ?? 0, bankAccount1?.hash ?? 0)
        XCTAssertEqual(bankAccount1?.hash ?? 0, bankAccount2?.hash ?? 0)
    }

    // MARK: - Description Tests

    func testDescription() {
        let bankAccount = STPBankAccount.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("BankAccount"))
        XCTAssert(bankAccount?.description)
    }

    // MARK: - STPAPIResponseDecodable Tests

    func testDecodedObjectFromAPIResponseRequiredFields() {
        let requiredFields = [
            "id",
            "last4",
            "bank_name",
            "country",
            "currency",
            "status",
        ]

        for field in requiredFields {
            var response = STPTestUtils.jsonNamed("BankAccount")
            response?.removeValue(forKey: field)

            XCTAssertNil(STPBankAccount.decodedObject(fromAPIResponse: response))
        }

        XCTAssert(STPBankAccount.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("BankAccount")))
    }

    func testDecodedObjectFromAPIResponseMapping() {
        let response = STPTestUtils.jsonNamed("BankAccount")
        let bankAccount = STPBankAccount.decodedObject(fromAPIResponse: response)

        XCTAssertEqual(bankAccount?.stripeID, "ba_1AZmya2eZvKYlo2CQzt7Fwnz")
        XCTAssertEqual(bankAccount?.accountHolderName, "Jane Austen")
        XCTAssertEqual(bankAccount?.accountHolderType ?? 0, Int(STPBankAccountHolderTypeIndividual))
        XCTAssertEqual(bankAccount?.bankName, "STRIPE TEST BANK")
        XCTAssertEqual(bankAccount?.country, "US")
        XCTAssertEqual(bankAccount?.currency, "usd")
        XCTAssertEqual(bankAccount?.fingerprint, "1JWtPxqbdX5Gamtc")
        XCTAssertEqual(bankAccount?.last4, "6789")
        XCTAssertEqual(bankAccount?.routingNumber, "110000000")
        XCTAssertEqual(bankAccount?.status ?? 0, Int(STPBankAccountStatusNew))

        XCTAssertNotEqual(bankAccount?.allResponseFields, response)
        XCTAssertEqual(bankAccount?.allResponseFields, response)
    }
}